### PR TITLE
KODA V2 & token swap

### PR DESCRIPTION
**Pull request by jamesmorgan**
* [x] New KODA V2 contract
* [x] Token Swap contract for V1 to V2
* [x] Migration tools - see https://github.com/knownorigin/token-swap-v2
* [x] Rebuild webapp with new contract integration

### CODE REVIEW

File: contracts/v2/KnownOriginDigitalAssetV2.sol [hunk 1]
```diff
@@ -0,0 +1,500 @@
...
+    // TODO method checking active (dates)
+    uint32 auctionStartDate;
+    uint32 auctionEndDate;
+
+    address artistAccount; // TODO duplicated between editions
+    uint256 priceInWei;
+    string tokenURI;
+
+    // Counters
+    uint8 sold;
```
Comment (andygray) [inline contracts/v2/KnownOriginDigitalAssetV2.sol]: I'd use `minted` as all may not have been sold...maybe.

File: contracts/v2/KnownOriginDigitalAssetV2.sol [hunk 2]
```diff
@@ -0,0 +1,500 @@
...
+    uint8 available;
+    bool active;
+  }
+
+  mapping(uint256 => EditionDetails) internal editionNumberToEditionDetails;
+
+  mapping(uint256 => uint256) internal tokenIdToEditionNumber;
+
+  mapping(uint256 => uint256[]) internal editionNumberToTokenIds;
```
Comment (andygray) [inline contracts/v2/KnownOriginDigitalAssetV2.sol]: this could be useful

File: contracts/v2/KnownOriginDigitalAssetV2.sol [hunk 3]
```diff
@@ -0,0 +1,500 @@
...
+    bool afterStartDate = editionNumberToEditionDetails[_editionNumber].auctionStartDate >= block.timestamp;
+    bool beforeStartDate = editionNumberToEditionDetails[_editionNumber].auctionEndDate <= block.timestamp;
+    require(afterStartDate && beforeStartDate);
+    _;
+  }
+
+  /*
+   * Constructor
+   */
+  constructor () public ERC721Token("KnownOriginDigitalAsset", "KODA") {
```
Comment (andygray) [inline contracts/v2/KnownOriginDigitalAssetV2.sol]: use `k0DA` to differentiate from v1?

File: contracts/v2/KnownOriginDigitalAssetV2.sol [hunk 4]
```diff
@@ -0,0 +1,500 @@
...
+  ///////////////
+  // Modifiers //
+  ///////////////
+
+  modifier onlyKnownOrigin() {
+    require(whitelist(msg.sender));
+    _;
+  }
+
+  modifier onlyEditionNotSoldOut(uint256 _editionNumber) {
```
Comment (andygray) [inline contracts/v2/KnownOriginDigitalAssetV2.sol]: `onlyAvailableEdition`?

Comment (andygray) [reply contracts/v2/KnownOriginDigitalAssetV2.sol]: `onlyNotSoldOutEdition`

File: contracts/v2/KnownOriginDigitalAssetV2.sol [hunk 5]
```diff
@@ -0,0 +1,500 @@
...
+    uint8 _available
+  )
+  public
+  onlyKnownOrigin
+  returns (bool)
+  {
+    // TODO validation
+
+    uint32 auctionEndDate = MAX_UINT32;
+    if(_auctionEndDate != 0){
```
Comment (andygray) [inline contracts/v2/KnownOriginDigitalAssetV2.sol]: if(_auctionEndDate == 0){

File: contracts/v2/KnownOriginDigitalAssetV2.sol [hunk 6]
```diff
@@ -0,0 +1,500 @@
...
+//  onlyActiveEdition(_editionNumber) // TODO is this correct to enforce this?
+//  onlyAfterPurchaseFromTime(_editionNumber) // TODO is this correct to enforce this?
+  returns (uint256) {
+
+    EditionDetails storage _editionDetails = editionNumberToEditionDetails[_editionNumber];
+
+    require(msg.value >= _editionDetails.priceInWei);
+
+    // Bump number sold
+    _editionDetails.sold = _editionDetails.sold + 1;
```
Comment (andygray) [inline contracts/v2/KnownOriginDigitalAssetV2.sol]: use `safeMath` : `_editionDetails.sold.add(1)`

File: contracts/v2/KnownOriginDigitalAssetV2.sol [hunk 7]
```diff
@@ -0,0 +1,500 @@
...
+
+    // Construct next token ID
+    uint256 _tokenId = _editionDetails.editionNumber + _editionDetails.sold;
```
Comment (andygray) [inline contracts/v2/KnownOriginDigitalAssetV2.sol]: `safeMath`

File: contracts/v1/ERC721BasicToken.sol [hunk 8]
```diff
@@ -2,8 +2,8 @@ pragma solidity ^0.4.21;
 
 import "./ERC721Basic.sol";
 import "./ERC721Receiver.sol";
-import "zeppelin-solidity/contracts/math/SafeMath.sol";
-import "zeppelin-solidity/contracts/AddressUtils.sol";
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
+import "openzeppelin-solidity/contracts/AddressUtils.sol";
```
Comment (andygray) [inline contracts/v1/ERC721BasicToken.sol]: 👍

File: test/helpers/makeInterfaceId.js [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+      }
+      return memo;
+    }, Buffer.alloc(INTERFACE_ID_LENGTH));
+
+  return `0x${interfaceIdBuffer.toString('hex')}`;
+}
+
+module.exports = {
+  makeInterfaceId,
+};
```
Comment (andygray) [inline test/helpers/makeInterfaceId.js]: did you make this or copy form somewhere - useful

File: test/truffle/v2/KnownOriginDigitalAssetV2.ERC721.test.js [hunk 10]
```diff
@@ -0,0 +1,911 @@
...
+
+      it('has a symbol', async function () {
+        const tokenSymbol = await this.token.symbol();
+        tokenSymbol.should.be.equal(symbol);
+      });
+
+      it('sets and returns metadata for a token id', async function () {
+        await this.token.setTokenURI(firstTokenId, sampleUri);
+        const uri = await this.token.tokenURI(firstTokenId);
+        uri.should.be.equal(`https://ipfs.infura.io/ipfs/${sampleUri}`);
```
Comment (andygray) [inline test/truffle/v2/KnownOriginDigitalAssetV2.ERC721.test.js]: could read the baseUrl too from contract in setup. Minor.

File: test/truffle/v2/KnownOriginDigitalAssetV2.ERC721.test.js [hunk 11]
```diff
@@ -0,0 +1,911 @@
...
+              // log.args._operator.should.be.equal(owner);
+              // log.args._from.should.be.equal(owner);
+              // log.args._tokenId.toNumber().should.be.equal(tokenId);
+              // log.args._data.should.be.equal(data);
+            });
+
+            it('should call onERC721Received from approved', async function () {
+              const result = await transferFun.call(this, owner, this.to, tokenId, {from: approved});
+              result.receipt.logs.length.should.be.equal(2);
+              // const [log] = decodeLogs([result.receipt.logs[1]], ERC721Receiver, this.receiver.address);
```
Comment (andygray) [inline test/truffle/v2/KnownOriginDigitalAssetV2.ERC721.test.js]: add //FIXME

File: test/truffle/v2/KnownOriginDigitalAssetV2.custom.test.js [hunk 12]
```diff
@@ -0,0 +1,1110 @@
...
+
+  const account1 = accounts[1];
+  const account2 = accounts[2];
+  const account3 = accounts[4];
+  const account4 = accounts[5];
+
+  const ZERO_ADDRESS = '0x0000000000000000000000000000000000000000';
+
+  const artistAccount = accounts[8];
+  const artistCommission = 76;
```
Comment (andygray) [inline test/truffle/v2/KnownOriginDigitalAssetV2.custom.test.js]: this is really `artistShare` as we as KO get commission they get paid

File: test/truffle/v2/KnownOriginDigitalAssetV2.custom.test.js [hunk 13]
```diff
@@ -0,0 +1,1110 @@
...
+      await this.token.createEdition(editionNumber1, editionData1, editionType, 0, 0, artistAccount, artistCommission, edition1Price, editionTokenUri1, 3, {from: _owner});
+      await this.token.createEdition(editionNumber2, editionData2, editionType, 0, 0, artistAccount, artistCommission, edition2Price, editionTokenUri2, 4, {from: _owner});
+    });
+
+    describe('checking raw edition data on creation', function () {
+
+      it('edition 1 setup correctly', async function () {
+        let edition = await this.token.allEditionData(editionNumber1);
+
+        web3.toAscii(edition[0]).replace(/\0/g, '').should.be.equal(editionData1); //_editionData
```
Comment (andygray) [inline test/truffle/v2/KnownOriginDigitalAssetV2.custom.test.js]: we should build a helper for this - I am copying this all over the shop in `art-blocks`

Comment (andygray) [reply test/truffle/v2/KnownOriginDigitalAssetV2.custom.test.js]: we really need to extract helpers and common utils into a new package/repo - TODO

File: test/truffle/v2/KnownOriginDigitalAssetV2.custom.test.js [hunk 14]
```diff
@@ -0,0 +1,1110 @@
...
+        edition[9].should.be.bignumber.equal(1); // _available
+        edition[10].should.be.equal(false); // _active
+      });
+
+      it('editionActive', async function () {
+        let active = await this.token.editionActive(editionNumber3);
+        active.should.be.equal(false);
+      });
+
+      it('can be created by partners', async function () {
```
Comment (andygray) [inline test/truffle/v2/KnownOriginDigitalAssetV2.custom.test.js]: first impression is use another contract - keep simple

File: test/truffle/v2/KnownOriginDigitalAssetV2.custom.test.js [hunk 15]
```diff
@@ -0,0 +1,1110 @@
...
+        await assertRevert(this.token.tokenURI(99));
+      });
+
+      // TODO test this after a burn
+
+    });
+
+    describe('tokenURISafe', async function () {
+
+      it('will return the correct edition ', async function () {
```
Comment (andygray) [inline test/truffle/v2/KnownOriginDigitalAssetV2.custom.test.js]: is `tokenURISafe` needed - what would it show?

File: test/truffle/v2/KnownOriginDigitalAssetV2.custom.test.js [hunk 16]
```diff
@@ -0,0 +1,1110 @@
...
+
+      it('reverts if edition invalid', async function () {
+        await this.token.updateAvailable(editionNumber1, 0, {from: _owner});
+
+        // reverts as edition sat to zero available
+        await assertRevert(this.token.mint(editionNumber1, {from: account1, value: edition1Price}));
+      });
+
+      it('reverts if edition auction not started', async function () {
+        // TODO
```
Comment (andygray) [inline test/truffle/v2/KnownOriginDigitalAssetV2.custom.test.js]: I can do these.

File: test/truffle/v2/KnownOriginDigitalAssetV2.custom.test.js [hunk 17]
```diff
@@ -0,0 +1,1110 @@
...
+        account2GasFees = await getGasCosts(receiptAccount2);
+
+        // post balances
+        postAccount1Balance = await web3.eth.getBalance(account1);
+        postAccount2Balance = await web3.eth.getBalance(account2);
+        postKoAccountBalance = await web3.eth.getBalance(await this.token.koCommissionAccount());
+        postArtistAccountBalance = await web3.eth.getBalance(artistAccount);
+      });
+
+      // beforeEach(() => {
```
Comment (andygray) [inline test/truffle/v2/KnownOriginDigitalAssetV2.custom.test.js]: drop?

File: test/truffle/v2/KnownOriginDigitalAssetV2.custom.test.js [hunk 18]
```diff
@@ -0,0 +1,1110 @@
...
+        await this.token.mint(editionNumber1, {from: account1, value: edition1Price});
+        await this.token.mint(editionNumber2, {from: account2, value: edition2Price});
+      });
+
+      it('should revert is token ID not valid', async function () {
+        await assertRevert(this.token.tokenIdentificationData(tokenIdInvalid));
+      });
+
+      it(`token id [${tokenId1}]`, async function () {
+        let results = await this.token.tokenIdentificationData(tokenId1);
```
Comment (andygray) [inline test/truffle/v2/KnownOriginDigitalAssetV2.custom.test.js]: `tokenIdentificationData` is this the token or the edition - presume token. Potentially needs rename.

Comment (andygray) [reply test/truffle/v2/KnownOriginDigitalAssetV2.custom.test.js]: is it just the edition data of the token?

Comment (andygray) [reply test/truffle/v2/KnownOriginDigitalAssetV2.custom.test.js]: would this have the totalSupply?

File: test/truffle/v2/KnownOriginDigitalAssetV2.custom.test.js [hunk 19]
```diff
@@ -0,0 +1,1110 @@
...
+      });
+    });
+
+  });
+
+  describe('burn', async function () {
+
+  });
+
+});
```
Comment (andygray) [inline test/truffle/v2/KnownOriginDigitalAssetV2.custom.test.js]: EPIC as Isaac would say these days...

File: test/truffle/v2/KnownOriginDigitalAssetV2.custom.test.js (hunk 14 above)
Comment (jamesmorgan) [reply test/truffle/v2/KnownOriginDigitalAssetV2.custom.test.js]: Agree, will chop

File: test/truffle/v2/KnownOriginDigitalAssetV2.custom.test.js (hunk 12 above)
Comment (jamesmorgan) [reply test/truffle/v2/KnownOriginDigitalAssetV2.custom.test.js]: 👍

File: test/truffle/v2/KnownOriginDigitalAssetV2.ERC721.test.js (hunk 11 above)
Comment (jamesmorgan) [reply test/truffle/v2/KnownOriginDigitalAssetV2.ERC721.test.js]: :thu

File: test/truffle/v2/KnownOriginDigitalAssetV2.ERC721.test.js (hunk 10 above)
Comment (jamesmorgan) [reply test/truffle/v2/KnownOriginDigitalAssetV2.ERC721.test.js]: 👍

File: contracts-flat/FLAT-KnownOriginDigitalAssetV2.sol [hunk 20]
```diff
@@ -0,0 +1,1948 @@
...
+
+// Utils only
+
+
+// TODO consider Pausable - stop open purchases?
+
+contract KnownOriginDigitalAssetV2 is
+ERC721Token,
+Whitelist,
+HasNoEther
```
Comment (jamesmorgan) [inline contracts-flat/FLAT-KnownOriginDigitalAssetV2.sol]: Add test

File: contracts/v2/KnownOriginDigitalAssetV2.sol [hunk 21]
```diff
@@ -0,0 +1,769 @@
...
+  mapping(uint256 => uint256) internal tokenIdToEditionNumber;
+
+  mapping(uint256 => uint256[]) internal editionNumberToTokenIds;
+  mapping(uint256 => uint256) internal editionNumberToTokenIdIndex;
+
+  mapping(address => uint256[]) internal artistToEditionNumbers;
+  mapping(uint256 => uint256) internal editionNumberToArtistIndex;
+
+  mapping(uint8 => uint256[]) internal editionTypeToEditionNumber;
+  mapping(uint256 => uint256) internal editionNumberToTypeIndex;
```
Comment (andygray) [inline contracts/v2/KnownOriginDigitalAssetV2.sol]: Talking about the art gallery stuff the type really comes into play - worth having this so we can build different front-ends but with KODA v2 core contract behind

File: contracts/v2/KnownOriginDigitalAssetV2.sol [hunk 22]
```diff
@@ -0,0 +1,769 @@
...
+
+  // TODO master list of editions - on creation? - how to handle edition burns/inactive?
```
Comment (andygray) [inline contracts/v2/KnownOriginDigitalAssetV2.sol]: Maybe leave this to the front end? Most should be active anyhow? And would be tricky to keep removing and adding based on active flag - on the other hand if the index pattern works active editions is your gallery load...hmmm

File: contracts/v2/KnownOriginDigitalAssetV2.sol [hunk 23]
```diff
@@ -0,0 +1,769 @@
...
+
+  modifier onlyAvailableEdition(uint256 _editionNumber) {
+    require(editionNumberToEditionDetails[_editionNumber].minted < editionNumberToEditionDetails[_editionNumber].available, "No more editions left to purchase");
+    _;
+  }
+
+  modifier onlyEditionsWithTokensAvailableToMint(uint256 _editionNumber) {
+    require(tokensOfEdition(_editionNumber).length != editionNumberToEditionDetails[_editionNumber].available, "No more editions left to mint");
+    _;
+  }
```
Comment (andygray) [inline contracts/v2/KnownOriginDigitalAssetV2.sol]: Are both `onlyEditionsWithTokensAvailableToMint` and `onlyAvailableEdition` needed - don't they do the same thing?

File: contracts/v2/KnownOriginDigitalAssetV2.sol [hunk 24]
```diff
@@ -0,0 +1,769 @@
...
+
+  modifier onlyActiveEdition(uint256 _editionNumber) {
+    require(editionNumberToEditionDetails[_editionNumber].active, "Edition not active");
+    _;
+  }
+
+  modifier onlyValidEdition(uint256 _editionNumber) {
```
Comment (andygray) [inline contracts/v2/KnownOriginDigitalAssetV2.sol]: Again could this be dropped for a single `onlyValidEdition` function that covers all bases?

File: contracts/v2/KnownOriginDigitalAssetV2.sol [hunk 25]
```diff
@@ -0,0 +1,769 @@
...
+    // TODO this needs to change as we could set the available to zero - replace with better check
+    require(editionNumberToEditionDetails[_editionNumber].available > 0, "No more editions available to purchase");
+    _;
+  }
+
+  modifier onlyAfterPurchaseFromTime(uint256 _editionNumber) {
```
Comment (andygray) [inline contracts/v2/KnownOriginDigitalAssetV2.sol]: `onlyAfterPurchaseFromTime` >> `onlyPurchaseDuringWindow`?

File: contracts/v2/KnownOriginDigitalAssetV2.sol [hunk 26]
```diff
@@ -0,0 +1,769 @@
...
+  uint256 public totalNumberAvailable;
+
+  // Object for edition details
+  struct EditionDetails {
+    // Identifiers
+    uint256 editionNumber;    // the range e.g. 10000
+    bytes32 editionData;      // some data about the edition
+    uint8 editionType;        // e.g. 1 = KODA V1 physical, 2 = KODA V1 digital, 3 = KODA V2, 4 = KOTA
+    // Config
+    uint32 auctionStartDate;  // date when the asset goes on sale
```
Comment (andygray) [inline contracts/v2/KnownOriginDigitalAssetV2.sol]: not a auction - use `startDate` and `endDate`?

File: contracts/v2/KnownOriginDigitalAssetV2.sol [hunk 27]
```diff
@@ -0,0 +1,769 @@
...
+   * Constructor
+   */
+  constructor () public ERC721Token("KnownOriginDigitalAsset", "KODA") {
+    // Assume the commission account is the creator for now
+    koCommissionAccount = msg.sender;
+    // Whitelist owner
+    addAddressToWhitelist(msg.sender);
+  }
+
+  function createActiveFullEdition(
```
Comment (andygray) [inline contracts/v2/KnownOriginDigitalAssetV2.sol]: What's a "Full" Edition?

File: contracts/v2/KnownOriginDigitalAssetV2.sol [hunk 28]
```diff
@@ -0,0 +1,769 @@
...
+    uint256 _priceInWei, string _tokenURI, uint8 _available
+  )
+  public
+  onlyKnownOrigin
+  returns (bool)
+  {
+    return _createEdition(_editionNumber, _editionData, _editionType, _auctionStartDate, _auctionEndDate, _artistAccount, _artistCommission, _priceInWei, _tokenURI, _available, true);
+  }
+
+  function createDisabledEdition(
```
Comment (andygray) [inline contracts/v2/KnownOriginDigitalAssetV2.sol]: getting confused between DisabledFull and Disabled?

File: contracts/v2/KnownOriginDigitalAssetV2.sol [hunk 29]
```diff
@@ -0,0 +1,769 @@
...
+    // Config
+    uint32 auctionStartDate;  // date when the asset goes on sale
+    uint32 auctionEndDate;    // date when the asset is available until
+    address artistAccount;    // artists account
+    uint8 artistCommission;   // base commissions, could be overridden by parent contracts
+    uint256 priceInWei;       // base price for asset, could be overridden by parent contracts
+    string tokenURI;          // IPFS Hash only
+    // Counters
+    uint8 minted;             // Total purchases/minted
+    uint8 available;          // Number with edition
```
Comment (andygray) [inline contracts/v2/KnownOriginDigitalAssetV2.sol]: I am finding `available` and `active` to mean the same thing when skimming. Maybe `available` should be `totalSupply` or `totalAvailable` - I know I probably suggested this to be `available` but just spend 5 mins figuring this out. maybe `max`

File: contracts/v2/KnownOriginDigitalAssetV2.sol [hunk 30]
```diff
@@ -0,0 +1,769 @@
...
+    address _artistAccount, uint8 _artistCommission,
+    uint256 _priceInWei, string _tokenURI,
+    uint8 _minted, uint8 _available
+  )
+  public
+  onlyKnownOrigin
+  returns (bool)
+  {
+    _createEdition(_editionNumber, _editionData, _editionType, _auctionStartDate, _auctionEndDate, _artistAccount, _artistCommission, _priceInWei, _tokenURI, _available, false);
+    updateMinted(_editionNumber, _minted);
```
Comment (andygray) [inline contracts/v2/KnownOriginDigitalAssetV2.sol]: presuming this is going to be rare - maybe do in two calls and keep the `createEditions` simple?

File: contracts/v2/KnownOriginDigitalAssetV2.sol [hunk 31]
```diff
@@ -0,0 +1,769 @@
...
+
+    // prevent missing token URI
+    require(bytes(_tokenURI).length != 0, "Token URI is missing");
+
+    // prevent empty artists address
+    require(_artistAccount != address(0), "Artist account not provided");
+
+    // Prevent commission of greater than 100% and less than 0%
+    require(_artistCommission <= 100, "Artist commission cannot be greater than 100");
+    require(_artistCommission >= 0, "Artist commission cannot be less than zero");
```
Comment (andygray) [inline contracts/v2/KnownOriginDigitalAssetV2.sol]: combine into single require?

File: contracts/v2/KnownOriginDigitalAssetV2.sol [hunk 32]
```diff
@@ -0,0 +1,769 @@
...
+      minted : 0, // default to all available
+      available : _available,
+      active : _active
+      });
+
+    // Add to total available count
+    totalNumberAvailable = totalNumberAvailable.add(_available);
+
+    // Maintain two way mappings so we can query direct
+    // e.g.
```
Comment (andygray) [inline contracts/v2/KnownOriginDigitalAssetV2.sol]: remove comment?

File: contracts/v2/KnownOriginDigitalAssetV2.sol [hunk 33]
```diff
@@ -0,0 +1,769 @@
...
+    // /type - DONE
+    // /editionNumber- DONE
+
+    _updateArtistLookData(_artistAccount, _editionNumber);
+    _updateEditionTypeLookData(_editionType, _editionNumber);
+
+    return true;
+  }
+
+  function _updateEditionTypeLookData(uint8 _editionType, uint256 _editionNumber) internal {
```
Comment (andygray) [inline contracts/v2/KnownOriginDigitalAssetV2.sol]: Look > LookUp

File: contracts/v2/KnownOriginDigitalAssetV2.sol [hunk 34]
```diff
@@ -0,0 +1,769 @@
...
+  }
+
+  function _updateArtistLookData(address _artistAccount, uint256 _editionNumber) internal {
+    uint256 artistEditionIndex = artistToEditionNumbers[_artistAccount].length;
+    artistToEditionNumbers[_artistAccount].push(_editionNumber);
+    editionNumberToArtistIndex[_editionNumber] = artistEditionIndex;
+  }
+
+  // This is the main purchase method
+  function mint(uint256 _editionNumber)
```
Comment (andygray) [inline contracts/v2/KnownOriginDigitalAssetV2.sol]: `mint` > `purchase`

File: contracts/v2/KnownOriginDigitalAssetV2.sol [hunk 35]
```diff
@@ -0,0 +1,769 @@
...
+  public payable
+  onlyAvailableEdition(_editionNumber)
+  onlyValidEdition(_editionNumber)
+  onlyActiveEdition(_editionNumber)
```
Comment (andygray) [inline contracts/v2/KnownOriginDigitalAssetV2.sol]: do you need the modifiers here as they are on the `mintTo` function it calls?

File: contracts/v2/KnownOriginDigitalAssetV2.sol [hunk 36]
```diff
@@ -0,0 +1,769 @@
...
+    EditionDetails memory _editionDetails = editionNumberToEditionDetails[_editionNumber];
+    return _editionDetails.available.sub(_editionDetails.minted);
+  }
+
+  function editionActive(uint256 _editionNumber) public view returns (bool) {
+    EditionDetails memory _editionDetails = editionNumberToEditionDetails[_editionNumber];
+    return _editionDetails.active;
+  }
+
+  function purchaseDatesToken(uint256 _tokenId) public view returns (uint32 _auctionStartDate, uint32 _auctionEndDate) {
```
Comment (andygray) [inline contracts/v2/KnownOriginDigitalAssetV2.sol]: maybe use `purchaseStartDate` and `purchaseEndDate`

**Pull request merged**